pallet/src/lib.rs

```diff
@@ -103,35 +74,43 @@ pub mod pallet {
                 );
                 <Authorities<T>>::put(authorities);
             }
-
-            <SessionInfo<T>>::put(SessionChange {
-                session_id: 0,
-                authorities_changed: true,
-                created_at: <frame_system::Pallet<T>>::block_number(),
-                next_authorities: authorities.to_vec(),
-            })
         }
 
         pub(crate) fn update_authorities(authorities: &[T::AuthorityId]) {
             <Authorities<T>>::put(authorities);
         }
 
-        pub(crate) fn new_session(changed: bool, authorities: Vec<T::AuthorityId>) {
-            if let Some(old_session) = <SessionInfo<T>>::get() {
-                let current_block = <frame_system::Pallet<T>>::block_number();
-
-                <SessionInfo<T>>::put(SessionChange {
-                    session_id: old_session.session_id + 1,
-                    authorities_changed: changed,
-                    created_at: current_block,
-                    next_authorities: authorities,
-                });
+        pub fn current_session() -> AuthoritySession<T::AuthorityId, T::BlockNumber> {
+            AuthoritySession {
+                session_id: Session::<T>::session_index(),
+                authorities: Self::authorities(),
+                // System::<T>::block_number() is a valid block number, so unwrap should always be ok here
+                stop_h: Session::<T>::estimate_next_new_session(System::<T>::block_number())
+                    .0
+                    .unwrap(),
             }
         }
 
-        pub(crate) fn deposit_log(change: AuthoritiesLog<T::AuthorityId, T::BlockNumber>) {
-            let log: DigestItem<T::Hash> = DigestItem::Consensus(ALEPH_ENGINE_ID, change.encode());
-            <frame_system::Pallet<T>>::deposit_log(log);
+        pub fn next_session(
+        ) -> Result<AuthoritySession<T::AuthorityId, T::BlockNumber>, AlephApiError> {
+            Session::<T>::queued_keys()
+                .iter()
+                .map(|(_, key)| key.get(T::AuthorityId::ID).ok_or(AlephApiError::DecodeKey))
+                .collect::<Result<Vec<T::AuthorityId>, AlephApiError>>()
+                .and_then(|authorities| {
+                    Ok(AuthoritySession {
+                        session_id: Session::<T>::session_index() + 1,
+                        authorities,
+                        // System::<T>::block_number() is a valid block number, so unwrap should always be ok here
+                        stop_h: Session::<T>::estimate_next_new_session(
+                            Session::<T>::estimate_next_new_session(System::<T>::block_number())
```

Comment:
Could you define a tmp var for this with some name indicating what's going on? Like `next_session_start`.